api/dataprocessor_test.go

```diff
@@ -1095,6 +1096,194 @@ func TestMergeSeries(t *testing.T) {
+		srv.Cache.Add(metricKey, prevts, *itgen)
+		prevts = itgen.Ts()
+	}
+
+	// preparing the next itgen as store mock result (to simulate one that has just been added and not cached yet)
+	itgens := make([]chunk.IterGen, 0)
+	chunks[i].Series.Finish()
+	itgens = append(itgens, *chunk.NewBareIterGen(chunks[i].Series.Bytes(), chunks[i].Series.T0, chunkSpan))
+
+	steps := chunkSpan / 10
+	for from := uint32(chunkSpan + chunkSpan/2); from < uint32(chunkSpan*4); from += steps {
+		for to := from + 1; to <= from+chunkSpan+1; to += steps {
+			// resetting mock at the beginning of each iteration
+			store.ResetMock()
+			store.AddMockResult(itgens, nil)
+
+			req := reqRaw(metricKey, from, to, chunkSpan, 1, consolidation.None)
+			req.ArchInterval = 1
+			ctx := newRequestContext(&req, consolidation.None)
+			iters := srv.getSeriesCachedStore(ctx, to)
+
+			expectResFrom := from - (from % chunkSpan)
+			expectResTo := (to - 1) + (chunkSpan - (to-1)%chunkSpan) - 1
+			ts := make([]uint32, 0)
+			for _, it := range iters {
+				for it.Next() {
+					t, _ := it.Values()
+					ts = append(ts, t)
+				}
+			}
+			if ts[0] != expectResFrom {
+				t.Fatalf("From %d To %d; Expected first to be %d but got %d", from, to, expectResFrom, ts[0])
+			}
+			if ts[len(ts)-1] != expectResTo {
+				t.Fatalf("From %d To %d; Expected last to be %d but got %d", from, to, expectResTo, ts[len(ts)-1])
+			}
+		}
+	}
+}
+
+func TestGetSeriesAggMetrics(t *testing.T) {
+	cluster.Init("default", "test", time.Now())
+	stats, _ := helper.New(false, "", "standard", "metrictank", "")
+	store := mdata.NewMockStore()
+	chunkSpan := uint32(600)
+	numChunks := uint32(10)
+	metrics := mdata.NewAggMetrics(store, chunkSpan, numChunks, 0, 0, 0, 0, []mdata.AggSetting{})
+	mdata.InitMetrics(stats)
+	addr = "localhost:6060"
+	srv, _ := NewServer(stats)
+	srv.BindBackendStore(store)
+	srv.BindMemoryStore(metrics)
+	defer metrics.Stop()
+	from := uint32(1744)
+	to := uint32(1888)
+	metricKey := "metric1"
+	archInterval := uint32(10)
+	req := reqRaw(metricKey, from, to, 100, 10, consolidation.None)
+	req.ArchInterval = archInterval
+	ctx := newRequestContext(&req, consolidation.None)
+
+	metric := metrics.GetOrCreate(metricKey)
+	for i := uint32(50); i < 3000; i++ {
+		metric.Add(i, float64(i^2))
+	}
+
+	oldest, iters := srv.getSeriesAggMetrics(ctx)
+	timestamps := make([]uint32, 0)
+	values := make([]float64, 0)
+
+	for _, it := range iters {
+		for it.Next() {
+			ts, val := it.Values()
+			timestamps = append(timestamps, ts)
+			values = append(values, val)
+		}
+	}
+
+	// should be the T0 of the chunk from (1744) is in
+	// 1744 - (1744 % 600) = 1200
+	expected := uint32(1200)
+	if oldest != expected {
+		t.Errorf("Expected oldest to be %d but got %d", expected, oldest)
+	}
+
+	// number of returned ts should be the number of chunks the searched range spans across * chunkspan
+	// (1888 + (600 - 1888 % 600)) - (1744 - (1744 % 600)) = 1200
```

Comment:
I find this formula hard to follow.
we could just write
`// first chunk is 1200 - 1799 (600 values), second is 1800 - 2399 (600 values) -> 1200 values total`